feat(new source): Initial `exec` source

moogstuart opened the pull request:
Proposed fix for #992

Added exec source with streaming and scheduled capability.
Generates log events from stdout and stderr of a command execution.

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
feat(exec source): new `exec` source (#992) (#7)

Added exec source with streaming and scheduled capability.
Generates log events from stdout and stderr of a command execution.

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart commented:
I recently created a new exec source and was about to create an issue to cover this change when I noticed there was an existing issue that for this. Given that the existing issue is tagged with needs approval and rfc I am not sure if my pull request can be used but I’ll push and see. Looking over the original issue these are the obvious requirements that I can see and how the push request satisfies them (or not):-

> Before we begin work, we should outline scenarios this would support and how the incoming data should be structured, if at all. And would this support metrics?

Produces log events (similar to the stdin source)

> I think we could avoid forcing users to generate only one event per invocation (and thus each time necessarily create new process). The command could generate a stream of newline-delimited events until it exits.

Supports both scheduled and streaming modes. Also the scheduled mode can produce more than one event per execution if you have configured it to produce an event per line. 

> To improve reliability and flexibility we can add two additional parameter exit_restart_interval and which would specify the interval since the end of the previous invocation after which the command will be launched again.

The streaming mode currently does not support a restart interval.

> The one thing I'm a little hesitant about is encouraging long-running processes. Managing those correctly is non-trivial and not one of vector's core competencies. In the past, I've advocated for running those processes under a real process manager, connecting their stdout to a named pipe, and using the file source to ingest from the pipe.

The spawned process is setup to be killed on drop (child.kill_on_drop(true)). This might not satisfy all the long running process concerns but it seems to do a pretty good job of cleaning up after itself.

The current version does do anything with named pipes:

command.stdout(std::process::Stdio::piped()); command.stderr(std::process::Stdio::piped());

It includes a very basic ‘process manager’ for the streaming mode. Nothing fancy but it will respawn the process on exit (if desired) after a configurable interval.

> The ability to execute a script based collector on a cron-like schedule

Provides the ability to execute a scheduled or streaming command. The scheduled is currently simply based on a second interval.

> Anything sent to stdout by the script should be read as event data (line separated).

The streaming mode will create an event per stdout line.

The scheduled mode can create an event per stdout line or for the complete output of a single run.

> Anything send to stderr by the script should be internal logging to vector.

Both streaming and scheduled modes can process stderr if desired. The current version will create an event in the same way as for stdout (however the event will be tagged with certain fields to identify it e.g. stream or stdout or stderr, the pid, the command, exit code, the exec duration etc)

> The following exit codes should be interpreted by Vector: 0: successful completion 1: error in execution (vector should log stderr to internal logging)

The current version tags the event with the exit code (if running a scheduled command). A streaming command event will not include the exit code.

> Should have a limit for the execution time (the script author should be aware of that limit and has the responsibility to fail gracefully before timeout is reached).

The current version will timeout a scheduled command if it takes longer than the schedule interval to run.

> Should also (optionally) prevent concurrent execution (i.e. 1 minute schedule but 5 minute timeout).

The schedule timeout (based on the schedule interval) should prevent this.

> Should have options to nice and ionice the script being executed, where permissions allow.

No current functionality provided for this.

> Should be able to set directory for execution

The current_dir can be optionally set for both streaming and scheduled modes.

> Discord request - Is there an way to execute command "vcgencmd get_throttled" this command returns value. It's possible to run it in Vector and send results to Elastic-search ?

In conjunction with the elasticsearch sink the push request would satisfy this.

moogstuart commented:
Apologies, I did not realize the was a failing test:-

test sources::exec::tests::test_run_once_scheduled ... Mar 24 17:27:27.270 ERROR vector::sources::exec: Error during scheduled exec run.
Mar 24 17:27:27.270 ERROR vector::internal_events::exec: Unable to exec. error=Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }
FAILED

I will investigate and push a fix later today.

moogstuart committed:
fix(exec source) Disabled a broken windows test (#8)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
fix(exec source) Fixed some clippy warnings

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
fix(exec source) Fixed incorrectly including stderr (#10)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
fix(exec source) Fixed incorrectly including stderr p2 (#11)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

jszwedko commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,306 @@
+			}
+		}
+
+		pid_key: {
```
I realize there is some precedent for exposing these as configurable options to optionally add metadata, but since the introduction of remap, we've been moving more towards just setting the keys and then letting users use `remap` to rename them or delete them if they don't want them. What would you think of just always setting `pid`, `exit_status`, etc.?

moogstuart replied:
Agreed, it seems much cleaner to do it that way. I have now removed the optional keys and made them part of the log events all the time.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,868 @@
+    let output_string: String = String::from_utf8(output).unwrap();
+
+    if output_string.is_empty() {
+        None
```
I think we should still publish the event even if the output was empty.

moogstuart replied:
The logic is now slightly different with the streaming processing but the logic will now publish an event when a command finishes running with the exec_duration_millis, pid and exit_status (if available). This is essentially the equivalent of a command that runs and produces no data so I think this is satisfied.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,868 @@
+    output: Vec<u8>,
+    stream: &str,
+) -> Option<Vec<Event>> {
+    let output_string: String = String::from_utf8(output).unwrap();
```
We just treat these as bytes internally so I don't think you need to decode as UTF8 here just to convert to `Bytes` again later.

moogstuart replied:
The logic is different for the streaming processing (which we do for both now). There is conversion to utf8 as part of that processing but I think that is useful to ensure we are producing a valid utf8 string.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,868 @@
+                        }
+
+                        // Optionally include stderr
+                        if config.include_stderr.unwrap_or(true) {
```
What happens to `child.stderr` if nothing is reading from it (i.e. `include_stderr = false`? I'm wondering if it'll block.

moogstuart replied:
I have tested running a command that produced output on stdout and stderr. When I don't collect stderr everything still shutdown nicely.

jszwedko commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,306 @@
+		}
+		exec_interval_secs: {
+			common:        false
+			description:   "The interval in seconds between scheduled command runs."
```
I think we may want to highlight here that the command will be killed if it takes longer than `exec_interval_secs` to run.

moogstuart replied:
Added a comment as suggested.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,868 @@
+                    }
+                },
+                Err(_) => {
+                    error!("Timed out during scheduled exec run.");
```
The internal event should already emit a log so I don't think we need this one too.

moogstuart replied:
Removed

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,868 @@
+            }
+        }
+        Err(spawn_error) => {
+            error!("Error during scheduled exec run.");
```
Same here.

moogstuart replied:
Removed

jszwedko reviewed (commented):
This is great! Thank you so much @moogstuart . We've had a number of people ask for this feature and I'm sure they'll be happy to see it come to fruition.

I left some inline comments, but a few high level ones:

* I realize you already implemented a method for running processes continuously, but I share @lukesteensen's concerns from https://github.com/timberio/vector/issues/992 that this can be tricky and is not really Vector's core competency. Is this something you need for your use of vector? If so, could you expand on that a bit? Otherwise would you be ok with splitting that bit off and just implementing the scheduled process running for now? I think that covers what users have asked for up until now. We could revisit long running processes later.
* I'm a little concerned about buffering the output here when `event_per_line = true`. Would it be possible to handle it in a streaming fashion and push events onto the output channel `Pipeline` as lines come in? This would be similar to the way you implemented the streaming version.
* For `event_per_line = false` we may want to provide a config option to limit the upper bound to avoid blowing up vector if the command outputs a lot of text. `fluentbit` calls this `buf_size`.
* Would it be possible to add an integration test or two that would have the component run and actually exec an external process? The one for the `stdin` source might be a good example: https://github.com/timberio/vector/blob/5a35a2f1f4647a506187bc6850fa23b98f69f6f1/src/sources/stdin.rs#L160 . You could just start it up with an exec interval of 1s, wait 1s, and then assert the output.

jszwedko commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,306 @@
+			description: "The key name added to each event representing the arguments that were provided to the command."
+			required:    false
+			warnings: []
+			type: string: {
```
Should also be `[string]`

moogstuart replied:
Apologies but I think I missed one and now am confused. For the following should the type of the items in the array be string or [string]:-

>         arguments: {
>             common:      false
>             description: "Array of any arguments to pass to the command."
>             required:    false
>             type: array: {
>                 default: null
>                 items: type: [string]: {
>                     examples: ["Hello World!", "-la"]
>                     syntax: "literal"
>                 }
>             }
>         }

i.e. 
>                 items: type: [string]: {

or

>                 items: type: string: {

Sorry for the confusion!

jszwedko replied:
I think it should be `[string]` there and then the examples would become:

```
examples: [["Hello World"], ["-l", "-a"]]
```

Let me know if cue doesn't like that though.

jszwedko replied:
Hmm, actually, I'm wrong! I think what you have there is ok. I verified by looking at the `lua` cue which has a similar array argument.

moogstuart replied:
OK, I think the new documentation is correct now.

jszwedko commented on src/internal_events/exec.rs:
```diff
@@ -0,0 +1,51 @@
+use super::InternalEvent;
```
In addition to these metrics, I think we may want to publish metrics for:

* `events_in_total`, which was a standard metric added recently in https://github.com/timberio/vector/pull/6758
* `command_executed_total` which would be the total number of times the command was executed
* `command_execution_duration_ns` would would be a histogram of execution times

moogstuart replied:
Added

jszwedko commented on src/internal_events/exec.rs:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    fn emit_metrics(&self) {
+        counter!("processing_errors_total", 1);
```
Could we add `type = "failed"` and `exit_status` here? I think that will help people monitor failed executions.

moogstuart replied:
Add the type but not exit_status it was not available for a command spawn error. I have added exit_status to the command_executed_total so I hope that will be a reasonable substitute.

jszwedko commented on src/internal_events/exec.rs:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    fn emit_metrics(&self) {
+        counter!("processing_errors_total", 1);
```
Could we add `type = "timed_out"` here?

moogstuart replied:
Added

moogstuart committed:
refactor(exec source) Address review comments (#12)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
refactor(exec source) Address review comments (#13)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart commented:
> * I realize you already implemented a method for running processes continuously, but I share @lukesteensen's concerns from #992 that this can be tricky and is not really Vector's core competency. Is this something you need for your use of vector? If so, could you expand on that a bit? Otherwise would you be ok with splitting that bit off and just implementing the scheduled process running for now? I think that covers what users have asked for up until now. We could revisit long running processes later.

Regarding streaming mode. We would ideally like to keep the streaming mode but I'll remove it if it becomes a blocker. The reason we would like to have a streaming mode is we plan to install vector along side scripts (probably node scripts) that provide source capability that vector currently doesn't provide. Some of these script need to listen to sockets (Even though there is a socket source vector currently doesn't have additional functionality we would need e.g. SNMP) and continuously stopping and starting the process would interrupt the data flow. Also we do not want to install anything other than vector on the machines to stop/start the node process. I also think that once the capability to execute any command is provided then people could call anything anyway and the difference is only about how long vector allows the script to run. Someone could configure scheduled to run for a ‘long period of seconds’.

> * I'm a little concerned about buffering the output here when `event_per_line = true`. Would it be possible to handle it in a streaming fashion and push events onto the output channel `Pipeline` as lines come in? This would be similar to the way you implemented the streaming version.

I have modified the logic to handle the output of both scheduled and streaming commands in the same way. This was a great suggestion by the way and I think it really helped tidy up the logic.

> * For `event_per_line = false` we may want to provide a config option to limit the upper bound to avoid blowing up vector if the command outputs a lot of text. `fluentbit` calls this `buf_size`.

I have added an upper bound to the input buffer(s) called maximum_buffer_size. I set the default to be 1GB. There didn't seem to be a method I could use out of the box so I based the logic off the basic read method. Also, given that a maximum buffer size could chop a valid utf8 string in a place that makes it invalid so I had to add logic to handle that.

> * Would it be possible to add an integration test or two that would have the component run and actually exec an external process? The one for the `stdin` source might be a good example: https://github.com/timberio/vector/blob/5a35a2f1f4647a506187bc6850fa23b98f69f6f1/src/sources/stdin.rs#L160
>    . You could just start it up with an exec interval of 1s, wait 1s, and then assert the output.

I added an integration test called test_run_command_linux that executed a simple command and processed the output. The run_command function is the main function that is used by both scheduled and streaming modes.

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,306 @@
+		commonly_used: false
+		delivery:      "at_least_once"
+		deployment_roles: ["sidecar"]
+		development:   "stable"
```
I have changed it to beta.

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,306 @@
+			required:    false
+			type: array: {
+				default: null
+				items: type: string: {
```
Changed to [string]

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+            let duration = time::Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while shutdown is pending
+            while futures::poll!(shutdown.clone()).is_pending() {
+                let output = run_command(config.clone(), shutdown.clone(), out.clone()).await;
+
+                if let Err(command_error) = output {
+                    emit!(ExecFailed {
+                        command: config.command.as_str(),
+                        error: command_error,
+                    });
+                }
+
+                if futures::poll!(shutdown.clone()).is_ready() {
+                    break;
+                } else {
+                    warn!("Streaming processed ended before shutdown.");
+
+                    // Using time interval so it can utilize the shutdown signal
+                    let mut interval = time::interval(duration).take_until(shutdown.clone());
+
+                    // Call next twice since the first is immediate
+                    interval.next().await;
+                    interval.next().await;
+                }
+            }
```
I might use `tokio::select` here and do something like:

```rust
loop {
  tokio::select! {
     _ = &mut shutdown => break, // will break early if a shutdown is started
    output = run_command(config.clone(), shutdown.clone(), out.clone()) => {
      // handle command finished
    }
  }
}
```

Then when you want to sleep, instead of using `Interval`:

```
tokio::select! {
  _ = &mut shutdown => break, // will break early if a shutdown is started
  _ = sleep(duration) => {},
}
```

moogstuart replied:
It looks much cleaner, thank you for the suggestion. I have modified the logic to use `tokio::select!` (I had to clone the shutdown as it was used in multiple places).

jszwedko commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,237 @@
+				unit:    "bytes"
+			}
+		}
+		exec_interval_secs: {
```
We currently have a mix of config styles for these config options which are only relevant in specific "mode"s or "strategy"s. I think we are moving towards using nested maps for mode specific config, like the `aws_s3` source:

https://github.com/timberio/vector/blob/a7d3f047dbfc2ca2df1fa56ae187d504e8113c8a/docs/reference/components/sources/aws_s3.cue#L49-L124

What would you think of modifying this to move the mode specific config options under a key matching the mode (`scheduled` or `streaming`)? This would essentially be unflattening the config you flatten below.

moogstuart replied:
I have modified the logic and documentation so that the `scheduled` and `streaming` options appear nested under the relevant key e.g. `scheduled.exec_interval_secs`

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+            run_command(config.clone(), shutdown, tx),
+        );
+
+        let timeout_result = timeout.await;
+
+        match timeout_result {
+            Ok(output) => match output {
+                Ok(exit_status) => assert_eq!(0_i32, exit_status.unwrap().code().unwrap()),
+                Err(_) => panic!("Unable to run linux command"),
+            },
+            Err(_) => panic!("Timed out during test of run linux command."),
+        }
```
Nit: you could simplify this a bit like:

```rust
let output = timeout.expect("command timed out").expect("command errored")
```

moogstuart replied:
Fixed

jszwedko reviewed (commented):
Looking much cleaner, thanks @moogstuart ! I like how you refactored the scheduled and streaming task output handling.

I left some more in-line comments below.

I also still want to exercise this a bit locally by building to see if there are any missing edge cases. I'll aim to do that once the feedback is mostly settled.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+            let duration = time::Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while shutdown is pending
+            while futures::poll!(shutdown.clone()).is_pending() {
```
Can this and the one below use `poll!(&mut shutdown)` instead of cloning it each time? This is what the `generator` source does. The parameter would also need to be marked as `mut`, of course.

moogstuart replied:
This logic has now been changed to use `tokio::select!` per @jszwedko's suggestion. I tried to use `&mut shutdown` but with the new logic I could not think of a way to do that as `shutdown` was used in multiple places.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+                if futures::poll!(shutdown.clone()).is_ready() {
+                    break;
+                } else {
+                    warn!("Streaming processed ended before shutdown.");
```
```suggestion
                    warn!("Streaming process ended before shutdown.");
```

moogstuart replied:
Fixed

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+                    // Using time interval so it can utilize the shutdown signal
+                    let mut interval = time::interval(duration).take_until(shutdown.clone());
+
+                    // Call next twice since the first is immediate
+                    interval.next().await;
+                    interval.next().await;
```
Please elaborate here. Why is the first immediate? Could this use `interval_at` instead to eliminate the need for the first wait?

moogstuart replied:
I have changed this logic to use `tokio::select!` and `delay_for`. I think the first was immediate because of the underlying [interval](https://docs.rs/tokio/0.2.25/tokio/time/fn.interval.html) implementation, I mentions:-

> Creates new Interval that yields with interval of duration. The first tick completes immediately.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+    pub command: String,
+    pub arguments: Option<Vec<String>>,
+    pub current_dir: Option<PathBuf>,
+    pub include_stderr: Option<bool>,
```
Could this simply be a `bool` with a default to `true`, possibly with a `serde(skip_serializing_if = …)` if you don't want it to appear in default configs?

moogstuart replied:
Yes, thank you it is much better as a `bool`. I have changed it.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+
+    // Mark the start time just before spawning the process as
+    // this seems to be the best approximation of exec duration
+    let now = Instant::now();
```
Nit: I suggest naming this `start` to be clear to its purpose.

moogstuart replied:
Fixed

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+        byte_size: line.len(),
+    });
+
+    let mut event = Event::from(line);
+    let log_event = event.as_mut_log();
```
This is obsolete usage for new code. Use `let mut log_event = LogEvent::default();` and then return `log_event.into()` or `Event::Log(log_event)`

moogstuart replied:
Fixed. I explicitly added `message` and `timestamp`.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+fn buffer_to_string(buffer: &mut Vec<u8>, allow_shrinking: bool) -> Option<String> {
+    let mut left_over_buffer: Vec<u8> = Vec::new();
+    loop {
+        if let Ok(buffer_string) = String::from_utf8(buffer.clone()) {
```
`String::from_utf8` returns an error that can provide an indication of where the UTF-8 conversion failed (see [`Utf8Error::valid_up_to`](https://doc.rust-lang.org/std/str/struct.Utf8Error.html#method.valid_up_to)). I think this function could be rewritten with that in mind to avoid most of the overhead here.

moogstuart replied:
Thank you for pointing that out, I think that would be a really elegant way to implement this function. Per a suggestion from @jszwedko I have removed the UTF-8 logic for now. If you and @jszwedko would like be to add an `encoding` config option (e.g. binary or utf8 could be the initial options) I would be happy to add this in (and use your suggested `valid_up_to`).

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+
+        // Handle any left over buffer
+        if !buffer.is_empty() {
+            if let Some(buffer_string) = buffer_to_string(&mut buffer, true) {
+                let _ = sender.send((buffer_string, stream)).await;
+                if !buffer.is_empty() {
+                    info!("Invalid utf8, left in buffer.");
+                }
+            } else {
+                info!("Invalid utf8, left in buffer.");
+            }
```
This would probably be better with a `match`:
```suggestion
            match buffer_to_string(&mut buffer, true) {
                Some(buffer_string) => {
                    let _ = sender.send((buffer_string, stream)).await;
                    if !buffer.is_empty() {
                        info!("Invalid utf8, left in buffer.");
                    }
                }
                None => info!("Invalid utf8, left in buffer."),
            }
```

moogstuart replied:
Thank you. I will use this if we add the method back in. It is removed for now.

bruceg commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+            // by 3 and it is still invalid then assume the whole thing
+            // is invalid utf8. Don't shrink to smaller than 1 byte.
+            if allow_shrinking && left_over_buffer.len() < 3 && buffer.len() > 1 {
+                if let Some(last_byte) = buffer.pop() {
+                    left_over_buffer.insert(0, last_byte);
+                } else {
+                    buffer.append(&mut left_over_buffer);
+                    return None;
+                }
```
Ditto using a `match` here.

moogstuart replied:
Same as above

bruceg commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,237 @@
+				syntax: "literal"
+			}
+		}
+		command: {
+			required:    true
+			description: "The command to be run."
+			type: string: {
+				examples: ["echo", "./myscript.sh"]
+				syntax: "literal"
+			}
+		}
+		arguments: {
+			common:      false
+			description: "Array of any arguments to pass to the command."
+			required:    false
+			type: array: {
+				default: null
+				items: type: [string]: {
+					examples: ["Hello World!", "-la"]
+					syntax: "literal"
+				}
+			}
+		}
```
I was thinking to make the `command` parameter a list of strings just like `arguments` is now. Then the command is just `command[0]` and the arguments is `&command[1..]`.

Eventually it might be worth also accepting a single string. The common setup for executing that is to pass it through `$SHELL -c "COMMAND"` which then handles all the command-line parsing and setting up pipes or anything else. However, I don't think that should be the target for this initial feature.

moogstuart replied:
I like the idea of command being a list of strings as I am not sure how complex the parsing of arguments might be for different use cases but yes perhaps in the future it could be combined to a single string.

moogstuart replied:
I have now removed the arguments config item and changed command to an array.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+                respawn_on_exit: default_respawn_on_exit(),
+                respawn_interval_secs: default_respawn_interval_secs(),
+            },
+            command: "streamer".to_owned(),
```
This is only used in generating 'fake' events and not used to actually run a test streaming command. I have changed it to yes though.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+
+        // Not using the shutdown signal in this method so we need to shutdown
+        // in other methods to end the reading
+        while let Ok(bytes_read) = reader.read(&mut read_buffer).await {
```
It should be easy enough to change this to use lines() but before I change it can I clarify if we should also drop the maximum_buffer_size? I think if we use lines() or read_until \n then we cannot control the size of the string or bytes returned. I suppose we could chop the lines/bytes returned but it seems too late at that point.

jszwedko replied:
Right, if we switch to `lines()` we should drop `maximum_buffer_size` too. Given that the `stdin` source already has a precedent for just doing `lines()`, I'm ok with that for the initial implementation here. We could follow up later to re-add support for limiting the buffer size, which I think could be done with only slightly more complexity via `reader.by_ref().take(buf_size).read_until('\n', &mut read_buffer`.

moogstuart replied:
Oh that is really nice! I don't mind giving that implementation a shot and then switching to lines if I can't get it to work (or if you guys are unhappy with it). I really like how event_per_line and maximum_buffer_size work really well together for both streaming and scheduled so if we can get it to work using the method you suggested that would be great. If you want lines in order to keep the event_per_line applicable to both scheduled and streaming we could implement a poor mans maximum_buffer_size called maximum_lines. Does the above sound ok to you?

moogstuart replied:
I really liked your suggestion of using `take(buf_size)` and so I have implemented it. I had to `async_buf_read.rs` module as the `async_read.rs` one only implemented `AsyncRead` and not `AsyncBufRead` but other than that I think it turned out quite nicely.

Since you were OK with cutting off a UTF-8 character I removed the UTF validation logic and simply send the raw bytes. @bruceg suggested a really nice improvement to the `buffer_to_string` method using `Utf8Error::valid_up_to` so if we ever want to add some encoding logic back in we could conditionally add it back using the logic he suggested.

moogstuart committed:
refactor(exec source) Address review comments (#14)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,237 @@
+			}
+			exec_duration_millis: {
+				common:      false
+				description: "The duration in milliseconds a scheduled command took to complete."
```
The `exec_duration_millis` is sent on an event along with `exit_status` when the command finishes (this applies to both a scheduled command and a streaming command if it exits). This event is sent after the command finishes and so unlike the previous events it will not have `data_stream` set (as stdout/stderr is not relevant to the finishing of the command).

moogstuart commented on src/internal_events/exec.rs:
```diff
@@ -0,0 +1,130 @@
+                exit_status = "Unknown",
+                elapsed_millis = %self.exec_duration.as_millis(),
+            );
+        }
```
I added a method following the logic you suggested and use this within `emit_logs` and `emit_metric` logic.

moogstuart commented on src/internal_events/exec.rs:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    fn emit_metrics(&self) {
+        if self.exit_status.is_some() {
```
Fixed

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+}
+
+fn default_maximum_buffer_size() -> usize {
+    // 1GB
```
Whoops, thank you. Fixed.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+    #[serde(default = "default_maximum_buffer_size")]
+    pub maximum_buffer_size: usize,
+    #[serde(skip, default = "get_hostname")]
+    pub hostname: Option<String>,
```
Moved to the `build` method.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+pub const EXIT_STATUS_KEY: &str = "exit_status";
+pub const COMMAND_KEY: &str = "command";
+pub const ARGUMENTS_KEY: &str = "arguments";
+pub const EXEC_DURATION_MILLIS_KEY: &str = "exec_duration_millis";
```
They did not need to be public, I have corrected that thank you.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+        handle_exit_status(&config, pid, exit_status, elapsed, out).await;
+        Ok(Some(exit_status))
+    } else {
+        emit!(ExecCommandExecuted {
```
Captured

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+    command.stderr(std::process::Stdio::piped());
+
+    if let Some(arguments) = &config.arguments {
+        if !arguments.is_empty() {
```
It does not but I have now modified the logic as per @bruceg's suggestion to combine `command` and `arguments` into `command` so it is slightly different now.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+
+        let mut buffer: Vec<u8> = Vec::new();
+
+        let mut reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
```
Ok. I left it as is for now.

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+                counter += 1;
+            }
+
+            assert_eq!(8, counter);
```
Nice! Done

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,843 @@
+                counter += 1;
+            }
+
+            assert_eq!(9, counter);
```
Done

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,237 @@
+		current_dir: {
+			common:      false
+			required:    false
+			description: "The directory from within which to run the command."
```
Fixed

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,237 @@
+				description: "The process ID of the command."
+				required:    true
+				type: uint: {
+					examples: [0, 1]
```
Whoops! Good spot. I have modified the examples now.

moogstuart committed:
refactor(exec source) Merged with the latest master

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
refactor(exec source) Merged with the latest master (fixed clippy)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
Merge branch 'master' of https://github.com/timberio/vector into release

moogstuart commented:
Congratulation on the move to tokio 1.0!!! I have merged with the latest master and everything should be ready for your review after the break. Happy Easter!

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+        exec_duration,
+    });
+
+    let event = create_event(
```
I'm still on the fence about sending this final event since it doesn't actually correspond to a line from the output. Do you think it is necessary?

moogstuart replied:
My thought is that it is nice to sweep up the `exit_status` and `exec_duration_millis` and provide it in an event if we have it. It gives the downstream transformers something to act on if they want it. Just having the internal event (`ExecCommandExecuted`) might be ok but I was not sure. Conversely it might be a pain for downstream consumers to strip out if they are not interested. To be honest, I am really not sure. Happy to remove it and or keep it based on your final decision.

moogstuart replied:
Removed

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+
+        let mut read_buffer: Vec<u8> = Vec::new();
+
+        let reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
```
We can follow up on this later, but it might be nice to continue reading during the shutdown, with a time out. We would avoid executing the command anymore though (either scheduled or due to restart).

moogstuart replied:
Ok, I'll leave this for now though.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+            while let Ok(bytes_read) = limited_reader.read_until(b'\n', &mut read_buffer).await {
+                if bytes_read == 0 {
+                    // If we get a continuous stream of \n the bytes_read will be at least 1
+                    info!("End of input reached, stop reading.");
```
I'm wondering if this and some of the other logs wouldn't be better at the `debug` level.

moogstuart replied:
Agreed, I have changed them to debug.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+                if bytes_read == 0 {
+                    // If we get a continuous stream of \n the bytes_read will be at least 1
+                    info!("End of input reached, stop reading.");
+                    break;
```
We are already breaking here, so we can avoid the `else` below. Also,

moogstuart replied:
Fixed

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+                    break;
+                } else {
+                    // Strip of the end of line bytes
+                    if read_buffer.ends_with(&[b'\n']) {
```
I think the last line may not end with a `\n`. It may just end with an EOF.

moogstuart replied:
If the last line does not end with `\n` that is ok as this will only strip out the` \n` and` \r` if they are present. If not present it will still go on and create the event (same as if the buffer was filled up before the end of the line).

moogstuart replied:
Changed to use the codec now

jszwedko reviewed (commented):
Thanks @moogstuart ! This seems to be getting pretty close. I'll try to exercise it tomorrow.

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+				}
+			}
+		}
+		current_dir: {
```
Thanks for the review comments @jszwedko, I'll take a look at them later today but I wanted to comment on this one and one other first. I will change it to `working_directory` as I agree it is a better name. I generally like to stick to a ubiquitous language in all places which is why I stuck with `current_dir` as that is what the command named it.

moogstuart replied:
Done

moogstuart commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+            // Continue to loop while not shutdown
+            loop {
+                tokio::select! {
+                    _ = shutdown.clone() => break, // will break early if a shutdown is started
```
Even with `mut shutdown` I have been unable to get `&mut shutdown` to work in this method because of the combination of mutable and immutable borrows and the loop.

jszwedko replied:
Ah I see 🤔

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+						required:    false
+						warnings: []
+						type: uint: {
+							default: 60
```
Done, I have changed it to 5.

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+			required:    false
+			type: bool: default: true
+		}
+		maximum_buffer_size: {
```
Done

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+		platform_name: null
+	}
+
+	configuration: {
```
Since there are two different modes I am not sure how to do this elegantly. I could add a parallel enum (e.g. `ModeType`) similar to the current `Mode` just for specifying scheduled or streaming and then validate that it matches the `Mode` enum but that feels wrong. I could pull out the config from the enum and then match and validate on start up. I would love to keep the single combined enum and config but I can't think of a way to do that if the nested key can sometimes be scheduled and sometimes be streaming:-

`mode = scheduled
 scheduled.exec_interval_secs = 10`

or

`mode = scheduled
 streaming.respawn_on_exit = false`

I could do this but I am not sure that is what you want:-

`mode = scheduled
 mode_config.exec_interval_secs = 10`

or 

`mode = scheduled
 mode_config.respawn_on_exit = false`

Please let me know how you want to proceed with this (There might be a more elegant way but at the moment I am not seeing it).

jszwedko replied:
👍 My preference is for:

```
mode = scheduled
scheduled.exec_interval_secs = 10
```

Which is how the `aws_s3` source does it as well (given it only has one "mode" at the moment).

moogstuart replied:
Ok I will pull out the config from the Mode enum and have something like aws_s3 as follows:-

```
pub mode: Mode,
pub schedule: Option<ScheduledConfig>,
pub streaming: Option<StreamingConfig>,
```

I had been trying to avoid splitting out the config from the enum but I can't think of a way to do it.

moogstuart replied:
I created a issue request on the serde project related to this:- https://github.com/serde-rs/serde/issues/2013

moogstuart replied:
Changed to the new structure now.

moogstuart committed:
refactor(exec source) Address review comments (#19)

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

jszwedko commented on src/async_buf_read.rs:
```diff
@@ -0,0 +1,77 @@
+use tokio::io::{AsyncBufRead, AsyncRead, ReadBuf, Result as IoResult};
+
+pub trait VecAsyncBufReadExt: AsyncRead + AsyncBufRead {
+    /// Read data from this reader until the given future resolves.
```
```suggestion
    /// Allow reading data from this reader until the given future resolves.
```

moogstuart replied:
Corrected

jszwedko commented on src/async_buf_read.rs:
```diff
@@ -0,0 +1,77 @@
+};
+use tokio::io::{AsyncBufRead, AsyncRead, ReadBuf, Result as IoResult};
+
+pub trait VecAsyncBufReadExt: AsyncRead + AsyncBufRead {
```
👍 this is a nice wrapper. What would you think of adding a test or two to this module for it?

moogstuart replied:
I have added a test called `test_read_line_without_shutdown()` which is of some use I think. I would have liked to implement `test_read_line_with_shutdown` to test that it stops reading when the shutdown is triggered but I was having difficulty implementing it as I need to both read and write to it. I can give it another go as the review continues if you like.

moogstuart replied:
Removed it as it turned out wrappering the inner `AsyncRead` did the job. I did add some tests to `async_read.rs` if you want to keep them.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+                }
+
+                let read_bytes = Bytes::from(read_buffer.clone());
+                if sender.send((read_bytes, stream)).await.is_err() {
```
Under what conditions would this error? Should we be logging something here?

moogstuart replied:
I have added a log message. This will happen If the receive half of the channel is closed, either due to close being called or the Receiver handle dropping.

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+            }
+        } else {
+            // Keep reading max buffer chunks
+            while let Ok(bytes_read) = limited_reader.read_to_end(&mut read_buffer).await {
```
I think I see what you are doing here. I would have actually expected the buffer size to be a hard limit on the output for `event_per_line = false`. What would you think of just reading up until `buf_size` and just dropping the rest?

moogstuart replied:
I think because the logic is currently applying event_per_line to both scheduled and streaming then it should not be a hard limit but just a limit for each read/event. If you don't want to allow event_per_line for streaming then we could make it a hard limit. I kind of feel that the main purpose of the maximum_buffer_size_bytes is so we don't blow up memory on read which the limit per read/event should do (so we don't need to make it a hard limit for no more data). Actually even if event_per_line only applied to scheduled we might get a streaming process that malfunctions and starts spewing out something without a newline and I don't think we would want to drop the rest of its output. Let's discuss further if needed.

moogstuart replied:
Changed it to use the codec now which does throw away lines that are too long (I'm not sure I like that behavior but I figured it was ok and we could always replace the codec later if needed).

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+                // Clear the read buffer ready for the next read
+                read_buffer.clear();
+
+                // Reset the limit for the next read (minus any left over if we do utf8 checks)
```
```suggestion
                // Reset the limit for the next read
```

moogstuart replied:
Fixed

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,830 @@
+
+        let mut limited_reader = reader.take(buf_size);
+
+        if event_per_line {
```
Not a big deal, but it did strike me that maybe the below could be implemented using tokio codecs to unify the two branch arms.

If `event_per_line` is true, it could use https://docs.rs/tokio-util/0.6.5/tokio_util/codec/struct.LinesCodec.html
If `event_per_line` is false, it could use https://docs.rs/tokio-util/0.6.5/tokio_util/codec/struct.BytesCodec.html

moogstuart replied:
Cool! I was not aware of that. I'll need at bit more time to look at this tomorrow so I'll first submit the other fixes back to you.

jszwedko replied:
Aha, I think `event_per_line = true` should result in the `\r` and `\n` being trimmed.

moogstuart replied:
The `LinesCodec` does remove the end of line characters.

jszwedko replied:
This looks really good!

I think you could simplify it even further with something like:

```rust
let codec = if event_per_line {
  LinesCodec::new_with_max_length(buf_size)
} else {
  SizedBytesCodec::new_with_max_length(buf_size)
};
```

You might need to `Box` the codec since it is a trait object (`let codec: Box<dyn Codec>`).

That would allow you to DRY up the rest of this since it is identical aside from the codec creation.

jszwedko reviewed (commented):
Looking good @moogstuart ! Thanks for making all of the adjustments thus far.

I played around with this today. I left a few more inline comments, but also:

* if `event_per_line = false` could we add the exit status and duration metadata to the output event?
* could we call `data_stream` simply `stream`?
* the warning for "Streaming process ended before shutdown" seems to appear when the process is restarted rather than when it actually ended; could we have it appear when the process exits?
* I still think users are going to be tripped up having the final event output that doesn't correspond to a line but just has the exit status and duration. Would you be opposed to just dropping that for now? Or is it something you need?

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+			common:      false
+			description: "Include the output of stderr when generating events."
+			required:    false
+			type: bool: default: false
```
Fixed

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,231 @@
+		}
+		event_per_line: {
+			common:      false
+			description: "Determine if events should be generated per line."
```
I have modified the documentation text as you suggested. For a streaming process I think it is still possible to not want an `event_per_line` if your process outputs a continuous stream of bytes without a newline (e.g. `printf` without the `\n`). I agree that most of the time you would want `event_per_line` to be `true` though. I like the flexibility of being able to handle and with the default being true most of the time the user can ignore this config setting. Happy to move to only scheduled if you prefer though.

jszwedko replied:
Yeah, I guess I'm just wondering: if their stream has no newlines, how the events would otherwise be framed?

moogstuart replied:
Maybe fixed sized number of bytes maybe e.g. I tested with a script that used a `printf` to print hello every 3 seconds. I made the buffer size 5 and it worked perfectly. I do agree that this will be much less common though.

jszwedko replied:
Agreed, I think that will be fairly uncommon. We have been thinking about "codecs" recently though: https://github.com/timberio/vector/issues/6958. I think this could fit in with that work as a "fixed length" codec. Using the max buffer size to emulate that feels a bit indirect.

moogstuart replied:
Yes that sounds quite elegant! The tokio codecs are knew to me so I'll spend a bit of time trying to figure them out over the weekend.

moogstuart replied:
Changed to used the codec now. I really like it. I created a custom codec for bytes which we can remove and replace with the out of the box one if you want (I just liked the idea of applying the buffer to non-lined streamed output). I think if we go with the out of the box codec it will be of random size and the buffer size does not really apply.

moogstuart commented:
> * if `event_per_line = false` could we add the exit status and duration metadata to the output event?

Since we are streaming the stdout/stderr to events the exit_status and duration are not available (not until the process exits which in the steaming case is after we have received the data). We could add a duration the process has been running at the time we get the data but I suspect that will not be particularly useful.

> * could we call `data_stream` simply `stream`?

Done

> * the warning for "Streaming process ended before shutdown" seems to appear when the process is restarted rather than when it actually ended; could we have it appear when the process exits?

Fixed (it required an extra poll and shutdown.clone())

> * I still think users are going to be tripped up having the final event output that doesn't correspond to a line but just has the exit status and duration. Would you be opposed to just dropping that for now? Or is it something you need?

I think you are correct and most of the time people will not need it. I have removed it for now.

moogstuart committed:
refactor(exec source) Address review comments (#20)

* refactor(exec source) Address review comments

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

* refactor(exec source) Address review comments

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
refactor(exec source) Changed to codec

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart commented:
Once the latest push has finished building successfully it will be ready for re-review. Thanks for your review time!

moogstuart committed:
refactor(exec source) Changed to codec

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
Merge branch 'master' of https://github.com/timberio/vector into release

moogstuart committed:
refactor(exec source) Changed to codec and merged with latest master

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
Merge branch 'master' of https://github.com/timberio/vector into release

moogstuart commented:
Merged with the latest master

jszwedko commented on src/async_read.rs:
```diff
@@ -59,3 +59,81 @@ where
+    use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader, BufWriter};
+
+    #[tokio::test]
+    async fn test_read_line_without_shutdown() {
```
👍 thanks for these tests! I think we should leave them.

moogstuart replied:
Ok, cool

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,886 @@
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
```
Do we need `rename_all = "snake_case"` here and below? The options are already snake cased.

moogstuart replied:
Your right! I have removed it

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,886 @@
+    }
+}
+
+// TODO: Maybe split into a shared codec module
```
If you felt like it, you could move it into a separate file in `src/sources/exec/` and rename this one `src/sources/exec/mod.rs`. We could extract it later if we find a use for it outside of the `exec` source.

I think it would be nice to have some tests specifically for the codec.

moogstuart replied:
Created the separate file with tests

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,886 @@
+                    // Send what ever is left over
+                    Some(buf.split_to(buf.len()))
+                } else {
+                    // Nothing left ot send
```
```suggestion
                    // Nothing left to send
```

moogstuart replied:
Corrected typo

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,886 @@
+impl_generate_config_from_default!(ExecConfig);
+
+impl ExecConfig {
+    pub(self) fn validate(&self) -> Result<(), ExecConfigError> {
```
I think we more commonly leave off the visibility modifier than use `pub(self)`. They are analogous.

https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself

moogstuart replied:
Removed the visibility modifier

jszwedko commented on src/sources/exec.rs:
```diff
@@ -0,0 +1,886 @@
+        let expected_command_string = format!("{:?}", expected_command);
+        let command_string = format!("{:?}", command);
+
+        assert_eq!(expected_command_string, command_string);
```
I'm not super crazy about comparing debug display representations, but I see that `Command` doesn't implement `PartialEq` or `Eq`.

moogstuart replied:
Agreed (hopefully it will not turn into a flakey test)

jszwedko reviewed (commented):
This is looking even better, thanks @moogstuart !

I left a few more comments, but nothing major.

@bruceg would you want to give this another look too?

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,245 @@
+					unit:    "milliseconds"
+				}
+			}
+			exit_status: {
```
Cleaned up the documentation to remove the unused keys.

moogstuart commented:
> `let codec = if event_per_line {
>   LinesCodec::new_with_max_length(buf_size)
> } else {
>   SizedBytesCodec::new_with_max_length(buf_size)
> };`

I don't think I can do this because the result objects are different e.g.

`   type Item = String;
    type Error = LinesCodecError`

vs

`   type Item = BytesMut;
    type Error = io::Error;`

Please let me know if there is a way around this as it would be nice to have the same processing code for both codecs.

moogstuart committed:
refactor(exec source) Review comments

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart committed:
Merge branch 'master' of https://github.com/timberio/vector into release

moogstuart commented:
Review comments addressed. Ready for re-review.

jszwedko commented:
@moogstuart thanks for the updates!

And, hmm, yeah, I see, I'm not sure if there is a good way to box that either. @bruceg might have thoughts.

I'll leave this for his review, but otherwise I think this is good to go.

bruceg commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+
+async fn handle_exit_status(
+    config: &ExecConfig,
+    exit_status: Option<ExitStatus>,
+    exec_duration: Duration,
+) {
+    let exit_status = match exit_status {
+        Some(exit_status) => exit_status.code(),
+        None => None,
+    };
```
This may be simpler to have the single caller that passes `Some` handle this above, which would eliminate this match block.
`handle_exit_status(&config, Some(exit_status.code()), elapsed)`

moogstuart replied:
Converted as you suggested. The `handle_exit_status` method has become simpler and simpler over the reviews but I thought it was still useful to keep it as a placeholder for future logic if needed e.g. if we ever decide to add back in an event with the exit status and duration.

bruceg reviewed (approved):
Looks pretty good. I left some comments and questions about a few relatively minor tweaks.

moogstuart committed:
Merge branch 'master' of https://github.com/timberio/vector into release

moogstuart committed:
refactor(exec source) Address code review comments

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

moogstuart commented:
> Looks pretty good. I left some comments and questions about a few relatively minor tweaks.

Thanks for the review comments @bruceg. I have addressed all of them apart from the doc one as I was unable to get that past check-docs.

moogstuart commented on docs/reference/components/sources/exec.cue:
```diff
@@ -0,0 +1,227 @@
+			description: "The command to be run, plus any arguments required."
+			type: array: {
+				items: type: string: {
+					examples: ["echo", "Hello World!", "ls", "-la"]
```
I tried an array but got this error message:-

> components.sources.exec.configuration.command.type.array.items.type.string.examples.0: conflicting values string and ["echo","Hello World!"] (mismatched types string and list):

I tried putting examples in different places but it didn't seem to work. I'll try an experiment with it a bit more but I wanted to pushed the other changes first.

jszwedko replied:
Yeah, I think the issue here is that the examples there are examples of the array elements rather than examples of lists of elements. I was able to update the cue config to allow arrays to be specified as examples. I wasn't able to push to this branch though.

Could you apply this diff?

```diff
0cb7faabc - (HEAD -> release, origin/release) Allow specifying examples for config options that take an Array (4 minutes ago) <Jesse Szwedko>
diff --git a/docs/reference.cue b/docs/reference.cue
index d5e2f26b9..5bcaed70a 100644
--- a/docs/reference.cue
+++ b/docs/reference.cue
@@ -383,13 +383,18 @@ _values: {

 #TypeArray: {
        _args: required: bool
+       _type: items: type: string
+
        let Args = _args
+       let Type = _type

        if !Args.required {
                // `default` sets the default value.
                default: [...] | null
        }

+       examples?: [...[...Type.items.type]]
+
        // Set `required` to `true` to force disable defaults. Defaults should
        // be specified on the array level and not the type level.
        items: type: #TypePrimitive & {_args: required: true}
diff --git a/docs/reference/components/sources/exec.cue b/docs/reference/components/sources/exec.cue
index 944d76485..92cce4fa5 100644
--- a/docs/reference/components/sources/exec.cue
+++ b/docs/reference/components/sources/exec.cue
@@ -59,8 +59,8 @@ components: sources: exec: {
                        required:    true
                        description: "The command to be run, plus any arguments required."
                        type: array: {
+                               examples: [["echo", "Hello World!"], ["ls", "-la"]]
                                items: type: string: {
-                                       examples: ["echo", "Hello World!", "ls", "-la"]
                                        syntax: "literal"
                                }
                        }
```

moogstuart replied:
Done, thank you for the change.

moogstuart commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+
+// TODO: Would be nice to combine the scheduled and streaming config with the mode enum once
+//       this serde ticket has been addressed (https://github.com/serde-rs/serde/issues/2013)
+#[derive(Deserialize, Serialize, Debug, Clone)]
```
Added

moogstuart commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        self.validate()?;
+        let hostname = get_hostname();
+        match self.mode.clone() {
```
Added copy and changed this to &self.mode

moogstuart commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
```
I modified this as suggested and agree that it looks neater.

moogstuart commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+
+    // Mark the start time just before spawning the process as
+    // this seems to be the best approximation of exec duration
+    let start = Instant::now();
```
I removed the outer `Instant::now` as it was not needed. The scheduled command will timeout based on the schedule and so there is no need to monitor the time in addition to this.

moogstuart commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,798 @@
+
+    // Pipe our stdout/stderr to the process to we inherit it's output
+    command.stdout(std::process::Stdio::piped());
+    command.stderr(std::process::Stdio::piped());
```
I have added this to both stdin and stderr (when it is not being used).

moogstuart committed:
refactor(exec source) Applied document changes for array example

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

bruceg commented on src/sources/exec/mod.rs:
```diff
@@ -0,0 +1,789 @@
+    if config.include_stderr {
+        command.stderr(std::process::Stdio::piped());
+    } else {
+        command.stderr(std::process::Stdio::null());
```
I'm not sure this is wise, as it makes any error output invisible when we don't pull it in as input, but I'm not sure. @jszwedko?

jszwedko replied:
Hmm. In my ideal world:

* If `include_stderr` is `false` any stderr of the forked process is just dropped
* If `include_stderr` is `true` any stderr of the forked process is pushed as events

I think that is what is happening here, yes? I can test tomorrow.

bruceg replied:
Yes, that is what I would expect from this code. If that matches your expectations then I am fine with it.

moogstuart replied:
Just pushed a fix for the cue format (I am not sure why `cue fmt ./docs/**/*.cue` didn't fix it.

Regarding include_stderr, if you guys decide that this configuration is not needed (i.e. we would always need stderr) I would be happy to remove it. Perhaps stderr could be filtered out by a transformation instead.

jszwedko replied:
👍 I tested and the behavior seems to be what I expected as mentioned in the above comment.

moogstuart committed:
refactor(exec source) Fixed cue format issue

Signed-off-by: Stuart Broad <stuart@moogsoft.com>

jszwedko reviewed (commented):
Thanks @moogstuart !

I tried this out again today and it actually looks like the respawn behavior for streaming tasks isn't respecting the `streaming.respawn_interval_secs` interval. It seems to take 60s before it restarts my task even though I specified 5s:

```toml
[sources.in]
type = "exec"
streaming.respawn_on_exit = true
streaming.respawn_interval_secs = 5
command = ["ls", "-l", "/rcohuerch/"]
include_stderr = true

[sinks.out]
type = "console"
inputs = ["in"]
encoding.codec = "json"
```

The scheduled task bit looks great though!

moogstuart commented:
> Thanks @moogstuart !
> 
> I tried this out again today and it actually looks like the respawn behavior for streaming tasks isn't respecting the `streaming.respawn_interval_secs` interval. It seems to take 60s before it restarts my task even though I specified 5s:
> 
> ```toml
> [sources.in]
> type = "exec"
> streaming.respawn_on_exit = true
> streaming.respawn_interval_secs = 5
> command = ["ls", "-l", "/rcohuerch/"]
> include_stderr = true
> 
> [sinks.out]
> type = "console"
> inputs = ["in"]
> encoding.codec = "json"
> ```
> 
> The scheduled task bit looks great though!

Hi @jszwedko if you add `mode = "streaming"` it works as expected. The issue is that without specifying mode it defaults to `scheduled` and so uses the default scheduled duration.

jszwedko commented:
Aha, of course 😅 . Thanks @moogstuart . I think it'd be nice to warn for that case, but that is a nice-to-have. I think is good to go! Thanks so much for all of your hard work on it! I think this will be a popular source to fill gaps in vector's current source coverage and for custom inputs.

zsherman commented:
🥳
